pkg/ratelimiter/reconciler.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratelimiter
+
+import (
+	"context"
+	"time"
+
+	"sigs.k8s.io/controller-runtime/pkg/ratelimiter"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// A Reconciler rate limits an inner, wrapped Reconciler. Requests that are rate
+// limited immediately return RequeueAfter: d without calling the wrapped
+// Reconciler, where d is imposed by the rate limiter.
+type Reconciler struct {
+	name  string
+	inner reconcile.Reconciler
+	limit ratelimiter.RateLimiter
+}
+
+// NewReconciler wraps the supplied Reconciler, ensuring requests are passed to
+// it no more frequently than the supplied RateLimiter allows. Multiple uniquely
+// named Reconcilers can share the same RateLimiter.
+func NewReconciler(name string, r reconcile.Reconciler, l ratelimiter.RateLimiter) *Reconciler {
+	return &Reconciler{name: name, inner: r, limit: l}
+}
+
+// Reconcile the supplied request subject to rate limiting.
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	item := r.name + req.String()
+	if d := r.limit.When(item); d > time.Duration(0) {
+		return reconcile.Result{RequeueAfter: d}, nil
+	}
```

Comment:
Hrm, it actually looks like this doesn't work like I thought it worked.

I was thinking that `When` would return the duration after which calling `When` again would be successful (i.e. for the token bucket when it was refilled enough to handle new requests). In retrospect I believe what's happening under the hood (at least for a token bucket based rate limiter) is that it's taking a token from the bucket each time, then returning the time at which you're allowed to "cash in" the token you already took.

So in our case we shouldn't be calling `When` again in the `RequeueAfter` case and instead just proceeding, but unfortunately we don't have enough state for new reconciles to know whether they've been rate limited or not.

Reply:
Pushed a new commit that fixes this.